File: Medium/98. Validate Binary Search Tree/isValidBST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    /*
        根據BST的定義來解就行了.
        Time complexity: O(n)
        Space complexity: O(n)
    */
    bool isValidBST(TreeNode* root) {
        return isValid(root, LLONG_MIN, LLONG_MAX);
    }
private:
    bool isValid(TreeNode* root, long min_val, long max_val){
        if(!root){
           return true; 
        }
        if(root->val <= min_val || root->val >= max_val){
            return false;
        }
        return isValid(root->left, min_val, root->val) && isValid(root->right, root->val, max_val);
    }
};

File: Medium/98. Validate Binary Search Tree/isValidBST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return BST(root, Long.MAX_VALUE, Long.MIN_VALUE);
    }
    private boolean BST(TreeNode root, long max_v, long min_v){
        if(root == null){
            return true;
        }
        if(root.val <= min_v || root.val >= max_v){
            return false;
        }
        return BST(root.left, root.val, min_v) && BST(root.right, max_v, root.val);
    }
}